File: Top75_JS/33._Search_in_Rotated_Sorted_Array.js
// https://www.youtube.com/watch?v=A7dT6UtWEhA

function search(nums, target) {
  let left = 0;
  let right = nums.length - 1;

  while (left <= right) {
    let mid = Math.floor(left + (right - left) / 2);

    if (nums[mid] === target) {
      return mid;
    }
    if (nums[left] <= nums[mid]) {
      // Which means the left part is sorted
      if (nums[mid] > target && nums[left] <= target) right = mid - 1;
      else left = mid + 1;
    } else {
      if (nums[mid] < target && nums[right] >= target) left = mid + 1;
      else right = mid - 1;
    }
  }

  return -1;
}

File: Neetcode305/Binary_Search/33_Search_in_Rotated_Sorted_Array.py
class Solution:

    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1

        while left <= right:
            mid = (left + right) // 2

            if nums[mid] == target:
                return mid


# 我未成功的嘗試
#             if nums[mid]>=target and target > nums[right] :
# #                 左半邊，往右
#                 right=mid-1
#             elif nums[mid]<target and target < nums[left] :
# #                 右半邊，往左
#                 left=mid+1
#             elif nums[mid]>target and target < nums[right] :
#                 left=mid+1
# #               右半邊，往右
#             elif nums[mid]<target and target < nums[right] :
#                 right=mid-1

            if nums[mid] >= nums[left]:
                if target > nums[mid] or target < nums[left]:
                    left = mid + 1
                else:
                    right = mid - 1
            else:
                if target < nums[mid] or target > nums[right]:
                    right = mid - 1
                else:
                    left = mid + 1

        return -1